include/SFML/System/UniqueResource.hpp

```diff
@@ -0,0 +1,156 @@
+////////////////////////////////////////////////////////////
+//
+// SFML - Simple and Fast Multimedia Library
+// Copyright (C) 2007-2022 Laurent Gomila (laurent@sfml-dev.org)
+//
+// This software is provided 'as-is', without any express or implied warranty.
+// In no event will the authors be held liable for any damages arising from the use of this software.
+//
+// Permission is granted to anyone to use this software for any purpose,
+// including commercial applications, and to alter it and redistribute it freely,
+// subject to the following restrictions:
+//
+// 1. The origin of this software must not be misrepresented;
+//    you must not claim that you wrote the original software.
+//    If you use this software in a product, an acknowledgment
+//    in the product documentation would be appreciated but is not required.
+//
+// 2. Altered source versions must be plainly marked as such,
+//    and must not be misrepresented as being the original software.
+//
+// 3. This notice may not be removed or altered from any source distribution.
+//
+////////////////////////////////////////////////////////////
+
+#pragma once
+
+////////////////////////////////////////////////////////////
+// Headers
+////////////////////////////////////////////////////////////
+#include <SFML/System/Export.hpp>
+
+#include <optional>
+#include <utility>
+
+
+namespace sf
+{
+////////////////////////////////////////////////////////////
+/// \brief Class template for storing resources
+///        which cannot be copied
+///
+////////////////////////////////////////////////////////////
+template <typename T, typename Deleter>
+class UniqueResource
+{
+public:
+    ////////////////////////////////////////////////////////////
+    /// \brief Default constructor
+    ///
+    ////////////////////////////////////////////////////////////
+    UniqueResource();
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Construct from resource and deleter
+    ///
+    /// \param resource Resource to store
+    /// \param deleter Deleter for the resource
+    ///
+    ////////////////////////////////////////////////////////////
+    UniqueResource(T&& resource, Deleter&& deleter);
```

Comment:
Would it make sense to provide an explicit constructor that only accepts a `T&& resource` and default-initializes the deleter? Even the example in the doc would then not need a `{}` argument.

Or do you think it's a compatibility hazard toward `std::experimental::unique_resource`?

Reply:
I want there to be a clear path to `std::unique_resource` but I don't mind if it requires some minor changes like this.